project/src/js/ui/tournament/detail/tournamentDetailCtrl.js

```diff
@@ -24,33 +26,66 @@ export default function controller() {
     if (data.secondsToStart && data.secondsToStart > 0) {
       data.secondsToStart--;
     }
-
     if (data.secondsToFinish && data.secondsToFinish > 0) {
       data.secondsToFinish--;
     }
-
     m.redraw();
   }
 
+  function join(id) {
+    if (!id) {
+      id = tournament().id;
+    }
+    xhr.join(id).then(() => {
+      hasJoined(true);
+      m.redraw();
+    }, err => utils.handleXhrError(err));
+  }
+
+  function withdraw(id) {
+    if (!id) {
+      id = tournament().id;
+    }
+    xhr.withdraw(id).then(() => {
+      hasJoined(false);
+      m.redraw();
+    }, err => utils.handleXhrError(err));
+  }
+
+  let id = m.route.param('id');
+  let action = m.route.param('action');
```

Comment:
Ok, started doing this in a couple places. Will go through an make this change everywhere appropriate.